packages/apollo-server-defer/README.md

```diff
@@ -0,0 +1,173 @@
+## Proposal to add @defer support in Apollo Server
+
+### Motivation
+
+When loading data-heavy pages, we want to minimize the time to interactivity (TTI) in order to provide a snappy user experience. In many applications, we want TTI to be much shorter than how long the page would take to load completely - motivating the need for “above the fold” optimizations.
+
+In the world of REST APIs, a common solution is to make two separate requests to the server. The first loads minimal data to provide an initial render, and a second to load the rest of the data.
+
+In GraphQL/Apollo Client, using that same approach leads to the same developer overhead:
+
+- Figuring out how to chop up your query into different pieces and wrapping them with multiple query components
+- Sharing data between queries may be problematic (e.g. secondary queries might depend on some data returned from the first query)
+- Breaking up queries like this just doesn't feel like the smooth GraphQL experience we want 😢
+
+#### A Better Way
+
+_@defer_ provides a way for us to mark fields in our schema to _not_ wait on, because they might be:
+
+- _Expensive to load_. This includes private data that is not cached (like user progress), or information that requires more computation on the backend (like calculating price quotes on Airbnb)
+- _Not on the critical path for interactivity._ This includes loading the comments section of a blog post, or the number of claps received.
+- _Expensive to send._ Even if the field may resolve quickly (ready in memory), users might still choose to defer it if the cost of transport is too expensive.
+
+#### Use Cases
+
+```graphql
+type Video {
+  # public data that can be cached
+  id: ID!
+  title: String
+  description: String
+  # progress is private to the user, and significantly more
+  # expensive to retrieve
+  progress: Progress
+  comments: [Comment]
+}
+
+type Progress {
+  percent: Float
+  position: Float
+}
+
+type Comment {
+  userId: ID!
+  text: String
+}
+
+# Example Query using @defer
+query {
+  video(id: $id) {
+    title: String
+    progress @defer {
+      percent
+      position
+    }
+    comments @defer {
+      userId
+      text
+    }
+  }
+}
+```
+
+Instead of holding the GraphQL response and page rendering until the entire query is resolved, @defer tells Apollo Server to return a partial query response if the deferred fields are not ready yet. In the example above, we see how it may be used to remove `progress` and `comments` from the intial query response. Apollo Server would then take care of resolving the rest of the deferred fields in the background, and stream them to the client when they are ready.
+
+```
+// Initial Response
+{
+  "asset": {
+    "title": "Westworld",
```

Comment:
I've left some comments below, but I think we should return `progress` and `comments` as `null` here